cmd/trust/computepcr.go

```diff
@@ -0,0 +1,281 @@
+package main
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/canonical/go-efilib"
+	"github.com/canonical/tcglog-parser"
+	"github.com/urfave/cli"
+)
+
+const ShimLockGUID = "605dab50-e046-4300-abb6-3dd810dd8b23"
+const vendordbGUID = "00000000-0000-0000-0000-000000000000"
+const SBAT = "sbat,1,2021030218\012"
+
+
+var computePCR7Cmd = cli.Command{
+	Name:      "computePCR7",
+	Usage:     "Compute PCR7 value for a given keyset",
+	ArgsUsage: "<keyset-name>",
+	Action:    doComputePCR7,
+}
+
+func doComputePCR7(ctx *cli.Context) error {
+	args := ctx.Args()
+	if len(args) != 1 {
+		return errors.New("Required argument: keysetName")
+	}
+	keysetName := args[0]
+	if keysetName == "" {
+		return errors.New("Please specify a keyset name")
+	}
+	pcr7prod, pcr7lim, pcr7tpm, err := computepcr7(keysetName)
+	if err != nil {
+		return fmt.Errorf("Failed to generate pcr7 values for %s keyset: (%w)\n", keysetName, err)
+	}
+	fmt.Printf("uki-production: %x\nuki-limited: %x\nuki-tpm: %x\n", pcr7prod, pcr7lim, pcr7tpm)
+
+	return nil
+}
+
+type efiVarInfo struct {
+	varguid efi.GUID
+	hashed  []byte
+}
+
+func getCert(certfile string) ([]byte, error) {
+	// get size of certfile
+	info, err := os.Stat(certfile)
```

Comment:
It would be better to take the f that os.Open(certfile) returns, and call f.Stat() on that.  That way you can be sure the stat and open were done on the same file.  I'm not really worried about security implications of a TOCTTOU here, just about accidental crashes arising from them.

Reply:
ok, done.